Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp

```diff
@@ -1185,6 +1185,20 @@ int WebGLRenderingContextBase::drawingBufferHeight() const
     return m_context->getInternalFramebufferSize().height();
 }
 
+PredefinedColorSpace WebGLRenderingContextBase::drawingBufferColorSpace() const
+{
+    if (isContextLost())
+        return PredefinedColorSpace::SRGB;
+    return toPredefinedColorSpace(m_context->getDrawingBufferColorSpace()).value_or(PredefinedColorSpace::SRGB);
```

Comment:
I think this should just return m_drawingBufferColorSpace.

E.g. chain of logic:

All current implementations support all color spaces.

-> For all current implementations, setDrawingBufferColorSpace cannot fail.
-> if setDrawingBufferColorSpace can never fail, the get.. should not exist and we should cache this client-side.
-> Currently it is a bug to enable CanvasColorSpaceEnabled without the complete implementation.
-> Since the supported extra attributes are only P3, it's pointless to support "implementation" that doesn't support P3, as it then wouldn't support anything.

Reply:
I got rid of `WebGLRenderingContextBase::m_drawingBufferColorSpace`! I concurred with Dan suggesting that we should only have one source of truth -- so you think we should keep the set colorspace in both places?

> For all current implementations, setDrawingBufferColorSpace cannot fail.

Hmm, that's quite a different take that what I was working with!
So (to confirm) you're saying that whatever options exist in `PredefinedColorSpace` and/or(?) `DestinationColorSpace`, we can just assume that `setDrawingBufferColorSpace` will just deal with it, and the caller shouldn't care what really happens?
And from a JS test, `if ("drawingBufferColorSpace" in gl) { gl.drawingBufferColorSpace = "display-p3"; shouldBe(gl.drawingBufferColorSpace, "display-p3"); }` should always pass?

Reply:
> I got rid of `WebGLRenderingContextBase::m_drawingBufferColorSpace`! I concurred with Dan suggesting that we should only have one source of truth -- so you think we should keep the set colorspace in both places?

I think the way the code is currently written,`WebGLRenderingContextBase::m_drawingBufferColorSpace` is the only source of truth. 

Contrast:
- User selects clear color. We put it to m_clearColor and send it to the real context.
- User selects drawing buffer color space. We put it to m_drawingBufferColorSpace and send it to the real context.

All other things being equal, they're equal.

It makes no sense that drawing buffer color buffer would be anything else than what it was asked **if it cannot be nothing else**.  And it cannot be nothing else, if we don't have code for it to be nothing else.

> Hmm, that's quite a different take that what I was working with! So (to confirm) you're saying that whatever options exist in `PredefinedColorSpace` and/or(?) `DestinationColorSpace`, we can just assume that `setDrawingBufferColorSpace` will just deal with it, and the caller shouldn't care what really happens? And from a JS test, `if ("drawingBufferColorSpace" in gl) { gl.drawingBufferColorSpace = "display-p3"; shouldBe(gl.drawingBufferColorSpace, "display-p3"); }` should always pass?

The way the code is currently written, there is no error handling at all. If the client tries to put P3, this is propagated all the way to the IOSurface as is, and currently there is nothing else that is done or can be done.

This being the case, the simplest is to just .. write it that way explicitly.

So to recap:
- The feature should be enabled only when  PREDEFINED_COLOR_SPACE_DISPLAY_P3 is on and when the setting can be used.
- When the feature is enabled, gl.drawingBufferColorSpace = "display-p3";  selects P3.
- When P3 is selected, it is assumed to be selected. Because it **is** selected.

The situation becomes different if we have a consistent codepath which says "You tried to select P3. Currently due to some particular runtime configuration the selection could not be made, we did not respect it.". In this case, we should add the complexity of "select P3. Check if the selection succeeded or failed."

If we have such a codepath, we should also test it. E.g. write a test that ensures that P3 works. Then the test also toggles the test toggle that says "even if we support P3, manipulate the environment so we don't support P3." Then ensure that during this run, the P3 selection is not honoured.